rpcs3/util/vm_native.cpp

```diff
@@ -384,20 +391,30 @@ namespace utils
 			return false;
 		};
 
+		const std::string storage2 = fs::get_temp_dir() + "rpcs3_vm_sparse.tmp";
+		const std::string storage3 = fs::get_cache_dir() + "rpcs3_vm_sparse.tmp";
+
 		if (!storage.empty())
 		{
+			// Explicitly specified storage
 			ensure(f.open(storage, fs::read + fs::write + fs::create));
 		}
-		else if (!f.open(fs::get_cache_dir() + "rpcs3_vm", fs::read + fs::write + fs::create) || !set_sparse(f.get_handle(), m_size))
+		else if (!f.open(storage2, fs::read + fs::write + fs::create) || !set_sparse(f.get_handle(), m_size))
+		{
+			// Fallback storage
+			ensure(f.open(storage3, fs::read + fs::write + fs::create));
+		}
+		else
 		{
-			ensure(f.open(fs::get_temp_dir() + "rpcs3_vm", fs::read + fs::write + fs::create));
+			goto check;
 		}
```

Comment:
I don't like the go-to. Can't you just add a bool instead?

Reply:
I like neither, need to refactor this crap anyway to search for various locations where temp file could be placed.